Delete guest cart session if user logs in, solves #23686

tatusah opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

This code changes the session handler behaviour so that if a guest user who has products in cart later logs in the cart content is moved to the now logged in user user session and the guest session is deleted after login. Previously in woocommerce_sessions table would exist both the guest session and the logged in session. This would cause confusion as the same cart items would show up twice in calculations where one would check cart content from different sessions.

Closes #23686 .

### How to test the changes in this Pull Request:

1. Add products to cart while not logged in to user account. Check woocommerce_sessions table which shows the guest user session
2. Now log in to the user account
3. View the woocommerce_sessions table to see that now the guest session is removed and only logged in user session exists. Cart content has carried over correctly. Without this fix there would exist both the guest and logged in session.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Changelog entry

Changes the session handler behavior to delete the guest session after successful login.

tatusah committed:
Delete guest session after user logs in

Logging in to the user account makes a copy of the guest session and possible cart content. This causes conflicts in situations where there are only single quantities of products left and real time product availability checks are in place. While this kind of single quantity items are at time in two carts for the same visitor the checking of product availability fails. Deleting the same visitors guest cart solves this issue.

tatusah committed:
Update class-wc-session-handler.php

tatusah committed:
Removes the $old_session_key default value

Fix for the issue #23686

mikejolley commented:
Didn't this get handled in https://github.com/woocommerce/woocommerce/pull/23001 ?

tatusah committed:
Add mention of $old_session_key parameter

rekkitatu commented:
> Didn't this get handled in #23001 ?

To me it seems these are separate issues. The pull request you mentioned seemed to fix the sync issue between sessions opened on different browsers. The issue I'm trying to resolve is the left over guest sessions that exist in the session table after logging in. And these different left over sessions could be actually several per customer if the user would start adding products to cart as a guest on several devices and all of those would be left behind after logging in. The content of those would guest sessions would get merged to the logged in session but the guest sessions would still remain in the session table. until expired. At least in our case those left over sessions cause problems and getting rid of them would solve the issue.

This is my first time contributing so I'm a bit worried if I actually get the code completed correctly. But I'm sure I'll learn and get feedback from the more experienced around here.

mikejolley commented:
>  The issue I'm trying to resolve is the left over guest sessions that exist in the session table after logging in. And these different left over sessions could be actually several per customer if the user would start adding products to cart as a guest on several devices and all of those would be left behind after logging in. 

The PR I referenced migrates session data from the guest session to the user session after login. That was added in 3.6.

rekkitatu commented:
Without this fix there the woocommerce_session table would have two sessions for the same user, one with the hashed guest customer_id and another one with the actual logged in id and the same cart content would be in both of them like in the screenshot below. 

![image](https://user-images.githubusercontent.com/17417945/57623831-4a63a080-7599-11e9-8a60-104beb4eb908.png)

But do you foresee any ill effects that could be caused by deleting the quest sessions after the user logs in? Are those guest session details needed somewhere else after login?

rekkitatu commented:
> > The issue I'm trying to resolve is the left over guest sessions that exist in the session table after logging in. And these different left over sessions could be actually several per customer if the user would start adding products to cart as a guest on several devices and all of those would be left behind after logging in.
> 
> The PR I referenced migrates session data from the guest session to the user session after login. That was added in 3.6.

What do you mean exactly by migration (copying or overwriting)? Because to me it seems that WC 3.6.x functions so that during login a user session is created and the session data gets copied over from guest session to the newly created user session but the guest session still remains in database after that, doesn't it? I have tested this with the WC 3.6.2, storefront theme and no other plugins active besides just the Woocommerce. I would like to have the guest session deleted after migrating to the logged in user session.

tatusah committed:
Compatibility for persistent cart disabled case

If woocommerce_persistent_cart_enabled filter was set to false the cart would be emptied while logged in. This fixes also that case. Now the guest session is removed from the woocommerce_sessions table after successful login whether the persistent cart is set enabled or disabled and the guest session cart is carried over to logged in user session.

tatusah committed:
Add additional check for cart persistence

rekkitatu commented:
I have this code now running on production server. Working correctly, no issues.

pduclo commented:
I came across a glitch that this session handler fix causes.

Coupons don’t work for logged in users. They are fine for guests.

Coupons can be added by logged in users and appear ok at checkout, but are not applied. I tried three gateways. Paypal, stripe and cash on delivery.

Also, logged in users coupon disappears from checkout page if they change shipping options. Doesn’t happen to guests though.

I confirmed that reverting back to original sessionhandler file fixes this unintended issue.

rodrigoprimo commented on includes/class-wc-session-handler.php:
```diff
@@ -253,6 +256,10 @@ public function save_data() {
 
 			wp_cache_set( $this->get_cache_prefix() . $this->_customer_id, $this->_data, WC_SESSION_CACHE_GROUP, $this->_session_expiration - time() );
 			$this->_dirty = false;
+			$allow_guest_session_removal = apply_filters( 'woocommerce_persistent_cart_enabled', true ) ? true : (bool) get_user_meta( get_current_user_id(), '_woocommerce_load_saved_cart_after_login', true );
```
Could you please elaborate on why the guest session is only deleted if `woocommerce_persistent_cart_enabled` is `true` or if the user meta `_woocommerce_load_saved_cart_after_login` is `true`? It is not clear to me why this check is needed (I'm not saying it is not, just looking for clarification).

rekkitatu replied:
You are correct @rodrigoprimo. Thanks! I noticed that without those checks the cart didn't work correctly. However, it seems that by reviewing the code I was able to leave them out by changing the `$old_session_key` comparison to loose type. I made the final pull request and have tested out the this change with both the `woocommerce_persistent_cart_enabled` filter true & false and using both guest users and logged in users. Thanks for your help again.

rodrigoprimo reviewed (commented):
Thanks for your contribution, @rekkitatu.

I was able to reproduce the problem described in #23686, and I could confirm that this PR fixes it. I don't see a reason to keep the guest session in the database after the user logs in so what is prosed here seems fine to me (cc @woocommerce/proton for a second opinion).

I just left an inline comment asking for more clarification around the conditions when the code deletes the session.

tatusah committed:
Dropped redundant persistent cart check code

Changing the not equal check between current_user_id and old_session_key to loose type allowed leaving out persistent cart related check.

rodrigoprimo reviewed (approved):
This PR looks good to me. Leaving it open to get a second opinion from someone from @woocommerce/proton about deleting the guest session data.

rodrigoprimo reviewed (changes requested):
@rekkitatu, ignore my previous review. I just noticed that some unit tests are failing because they call WC_Session_Handler::save_data() without the first parameter. We need to either keep the first parameter optional or update the tests.

See: https://travis-ci.org/woocommerce/woocommerce/jobs/557528379#L656

peterfabian commented:
Since it's a `public` function, I think we need to keep the param optional.

peterfabian commented on includes/class-wc-session-handler.php:
```diff
@@ -253,6 +256,9 @@ public function save_data() {
 
 			wp_cache_set( $this->get_cache_prefix() . $this->_customer_id, $this->_data, WC_SESSION_CACHE_GROUP, $this->_session_expiration - time() );
 			$this->_dirty = false;
+			if ( get_current_user_id() != $old_session_key ) {
```
There is a slight problem I see with this approach. I think the `$guest_session_id` from above (`$old_session_key` in this context) is not always coming from the guest. If I'm logged in as User_1 and go to login screen and then log in as User_2 (without the logout step), session of User_1 gets deleted. 

Perhaps that's not a very common thing to do for users on an eshop, users could e.g. have a business account separate from their personal account and switch between those? 

I'm not sure about the best way to check this, though. Seems like the guest user's id is an MD5 hash, so maybe check if it has 32 characters and hope someone will fix it before an eshop can have 1x10^32 customers? :)

rekkitatu replied:
Oh, I didn't know it was possible to login as another user without logging out first. So I'll have to make the parameter optional and perhaps add check for the MD5 hash

```
function isValidMd5($md5 ='') {
  return strlen($md5) == 32 && ctype_xdigit($md5);
}
```

 or use WP function `get_user_by( 'id', $old_session_key )` to check if there is actual user related to the previous session ID as this should return false if the id passed is the guest user MD5 hash https://developer.wordpress.org/reference/functions/get_user_by/.

Do you @rodrigoprimo and @peterfabian think I should prefer one or the other way to check for case of changing from one logged in user to another one? MD5 check or the user object check?

rekkitatu replied:
Ended up checking for user object with the user ID / guest user hash.

peterfabian reviewed (changes requested):
I think it should work fine in most cases; I can see just one problem here with switching accounts without logging out.

tatusah committed:
Prevent conflict with multiple logged in users

Previous code would delete the session of previously logged in user in the rare case that multiple users would be logged in on the same browser. Added check to see if there is user account exists with the old_session_key and if not then the old session is from a guest user which can be deleted from database during logging in. Also returned old_session_key as optional parameter to prevent issues.

tatusah committed:
Coding standards fix

rekkitatu commented:
I wasn't able to figure out the cause for the Travis failure:

> 1) WC_Tests_WC_Product_Query::test_product_query_standard
Failed asserting that 0 matches expected 1.

Is this related to the code in the commit or is there just some problem with the test itself? Any tips @rodrigoprimo ?

kloon reviewed (approved):
Latest changes looks good to me. You can ignore the failing unit test, that is due to another PR that was merged.

Thanks for the contribution.
